src/openapi/streaming/subscription.ts

```diff
@@ -0,0 +1,1131 @@
+/* eslint-disable switch-case/no-case-curly */
+import { extend } from '../../utils/object';
+import log from '../../log';
+import {
+    ACTION_SUBSCRIBE,
+    ACTION_UNSUBSCRIBE,
+    ACTION_MODIFY_PATCH,
+    ACTION_UNSUBSCRIBE_BY_TAG_PENDING,
+} from './subscription-actions';
+import SubscriptionQueue from './subscription-queue';
+import type { QueuedItem } from './subscription-queue';
+import ParserFacade from './parser/parser-facade';
+import type { ITransport } from '../transport/trasportBase';
+import type { TransportCoreOptions } from '../transport/types';
+
+// TODO improve naming
+const updateTypes = {
+    UPDATE_TYPE_SNAPSHOT: 1,
+    UPDATE_TYPE_DELTA: 2,
+} as const;
+
+type SubscriptionUpdateTypes = typeof updateTypes[keyof typeof updateTypes];
+
+// TODO improve naming
+const stateFlags = {
+    SUBSCRIBE_REQUESTED: 1,
+    SUBSCRIBED: 2,
+    UNSUBSCRIBE_REQUESTED: 4,
+    UNSUBSCRIBED: 8,
+    PATCH_REQUESTED: 16,
+    READY_FOR_UNSUBSCRIBE_BY_TAG: 32,
+} as const;
+
+export type SubscriptionState = typeof stateFlags[keyof typeof stateFlags];
+
+export interface StreamingOptions {
+    headers?: Record<string, string>;
+    onUpdate?: (data: unknown, updateType: SubscriptionUpdateTypes) => void;
+    onError?: (data: unknown) => void;
+    onQueueEmpty?: () => void;
+    onNetworkError?: () => void;
+}
+
+export interface SubscriptionArgs {
+    Format?: string;
+    Arguments?: Record<string, unknown>;
+    RefreshRate?: number;
+    Top?: number;
+    Tag?: string;
+}
+
+interface SubscriptionSuccessResult {
+    State: 'active' | 'suspended';
+    Format: string;
+    ContextId: string;
+    InactivityTimeout: number;
+    RefreshRate: number;
+    Snapshot: Record<string, unknown>;
+    Schema?: string;
+    SchemaName?: string;
+}
+
+/**
+ * The static counter to generate unique reference id's.
+ */
+let referenceIdCounter = 1;
+
+const DEFAULT_REFRESH_RATE_MS = 1000;
+const MIN_REFRESH_RATE_MS = 100;
+
+const FORMAT_PROTOBUF = 'application/x-protobuf';
+const FORMAT_JSON = 'application/json';
+
+const ERROR_UNSUPPORTED_FORMAT = 'UnsupportedSubscriptionFormat';
+
+const LOG_AREA = 'Subscription';
+
+/**
+ * A subscription to a resource, which streams updates.
+ *
+ * This class should not be constructed directly, it should instead be created via the
+ * Streaming.createSubscription factory method.
+ */
+
+class Subscription {
+    UPDATE_TYPE_SNAPSHOT = 1 as const;
+    UPDATE_TYPE_DELTA = 2 as const;
+    // FIXME consider making those constants static - BREAKING CHANGE
+    STATE_SUBSCRIBE_REQUESTED = 1 as const;
+    STATE_SUBSCRIBED = 2 as const;
+    STATE_UNSUBSCRIBE_REQUESTED = 4 as const;
+    STATE_UNSUBSCRIBED = 8 as const;
+    STATE_PATCH_REQUESTED = 16 as const;
+    // FIXME this const was originally undefined - verify why
+    STATE_READY_FOR_UNSUBSCRIBE_BY_TAG = 32 as const;

```

Comment:
Please pay attention to this change and try to verify that I didn't make sth worse by introducing the missing flag.

Reply:
By adding the missing constant I think the behaviour is now change in [this line](https://github.com/hiredgun/openapi-clientlib-js/pull/28/files#diff-4efe8e9679cdce129f826267dd40340a26926acd7730c56c50d433cf86cbeee9R327) but I've found this PR and it seems like STATE_READY_FOR_UNSUBSCRIBE_BY_TAG was overlooked.

Reply:
May be you can check with owner of PR or Luke

Reply:
Talked to Luke, will leave my fix and will try to cover this case with unit tests in a separate PR